presto-main/src/main/java/com/facebook/presto/operator/HashAggregationOperator.java

```diff
@@ -360,20 +568,159 @@ protected Page computeNext()
                             BlockBuilder output = pageBuilder.getBlockBuilder(types.size() + i);
                             aggregator.evaluate(groupId, output);
                         }
-
-                        groupId++;
                     }
 
                     return pageBuilder.build();
                 }
             };
         }
+
+        private void rebuildAggregators()
+        {
+            groupByHash = createGroupByHash();
+            aggregators = createAggregators();
+        }
+
+        private void rebuildAggregatorsForMerge()
+        {
+            groupByHash = createGroupByHash();
+            aggregators = createAggregatorsForMerge(groupByHash.getTypes().size());
+        }
+
+        private GroupByHash createGroupByHash()
+        {
+            return GroupByHash.createGroupByHash(operatorContext.getSession(), groupByTypes, Ints.toArray(groupByChannels), hashChannel, expectedGroups);
+        }
+
+        private List<Aggregator> createAggregators()
+        {
+            // wrapper each function with an aggregator
+            ImmutableList.Builder<Aggregator> builder = ImmutableList.builder();
+
+            requireNonNull(accumulatorFactories, "accumulatorFactories is null");
+            for (int i = 0; i < accumulatorFactories.size(); i++) {
+                AccumulatorFactory accumulatorFactory = accumulatorFactories.get(i);
+                builder.add(new Aggregator(accumulatorFactory, step));
+            }
+            return builder.build();
+        }
+
+        private List<Aggregator> createAggregatorsForMerge(int numberOfColumnsInKey)
+        {
+            // wrapper each function with an aggregator
+            ImmutableList.Builder<Aggregator> builder = ImmutableList.builder();
+
+            requireNonNull(accumulatorFactories, "accumulatorFactories is null");
+            for (int i = 0; i < accumulatorFactories.size(); i++) {
+                AccumulatorFactory accumulatorFactory = accumulatorFactories.get(i);
+                builder.add(new Aggregator(accumulatorFactory, step, numberOfColumnsInKey + i));
+            }
+            return builder.build();
+        }
+
+        private void convertToMerge()
+        {
+            step = Step.partialInput(step);
+            ImmutableList.Builder<Integer> groupByPartialChannels = ImmutableList.builder();
+            for (int i = 0; i < groupByTypes.size(); i++) {
+                groupByPartialChannels.add(i);
+            }
+            if (hashChannel.isPresent()) {
+                hashChannel = Optional.of(groupByTypes.size());
+            }
+
+            groupByChannels = groupByPartialChannels.build();
+        }
+
+        private void closeSpiller()
+        {
+            if (spiller.isPresent()) {
+                spiller.get().close();
+                spiller = Optional.empty();
+            }
+        }
+
+        private List<Type> buildIntermediateTypes()
```

Comment:
yes, `intermediate` terminology comes from `GroupedAccumulator` and comes from `AggregationNode` but they more or less describe same thing.